[ao][sparsity] README for activation sparsifier

macandro96 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81814

The README contains introduction and details on the activation sparsifier. It also contains
code snippets and examples on using the activation sparsifier.

Test Plan: None

Commit:
[ao][sparsity] README for activation sparsifier

The README contains introduction and details on the activation sparsifier. It also contains
code snippets and examples on using the activation sparsifier.

Test Plan: None

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] README for activation sparsifier"


The README contains introduction and details on the activation sparsifier. It also contains
code snippets and examples on using the activation sparsifier.

Test Plan: None

[ghstack-poisoned]

z-a-f commented on torch/ao/sparsity/_experimental/activation_sparsifier/README.md:
```diff
@@ -0,0 +1,75 @@
+## How does it work?
+The idea is to compute a mask to prune the activations. To compute the mask, we need a representative tensor that generalizes activations coming from all the batches in the dataset.
+
+There are 3 main steps involved -
```
nit: `There are 3 main steps involved:`

z-a-f commented on torch/ao/sparsity/_experimental/activation_sparsifier/README.md:
````diff
@@ -0,0 +1,75 @@
+Essentially, the high level idea of computing the mask is
+
+```
+aggregated_tensor = aggregate_fn([activation for activation in all_activations])
+mask = mask_fn(reduce_fn(aggregated_tensor))
````
Split into two lines

z-a-f replied:
nit: For simple comments, consider using the "interactive python formatting":

```
>>> aggregated_tensor = aggregate_fn([activation for activation in all_activations])
>>> mask = mask_fn(reduce_fn(aggregated_tensor))
```

z-a-f commented on torch/ao/sparsity/_experimental/activation_sparsifier/README.md:
````diff
@@ -0,0 +1,75 @@
+mask = mask_fn(reduce_fn(aggregated_tensor))
+```
+
+*The activation sparsifier also supports per-feature/channel sparsity. This means that a desired set of features in an activation can be also pruned. The mask will be stored per feature.*
+
+```
+for i in range(len(features)):
+    aggregated_tensor_feature = aggregate_fn([activation[features[i]] for activation in all_activations])
+    mask[i] = mask_fn(reduce_fn(aggregated_tensor_feature))
+```
````
I think this is a little confusing. Could you elaborate more on this? Maybe give a comparative example with basic results or shapes?

macandro96 replied:
Added the difference in shape

z-a-f commented on torch/ao/sparsity/_experimental/activation_sparsifier/README.md:
````diff
@@ -0,0 +1,75 @@
+```
+
+## Implementation Details
+The activation sparsifier attaches itself to a set of layers in a model and then attempts to sparsify the activations flowing through them. *Attach* means registering a [```forward_pre_hook()```](https://pytorch.org/docs/stable/_modules/torch/nn/modules/module.html#register_forward_pre_hook) to the layer.
````
The inline "verbatim" code in markdown uses a single tick (not triple). For example:

**Inline:** "This is an \`inline piece of code\`"  =>  "This is an `inline piece of code`"
**Block:**
\`\`\`
This is a block piece of code
\`\`\`

=> 

```
This is a block piece of code
```

z-a-f commented on torch/ao/sparsity/_experimental/activation_sparsifier/README.md:
````diff
@@ -0,0 +1,75 @@
+```squash_mask```: Unregisters aggregate hook that was applied earlier and registers sparsification hooks if ```attach_sparsify_hook=True```. Sparsification hooks applies the computed mask to the activations before it flows into the registered layer.
+
+## Example
+```
````
nit: blank line before block

z-a-f commented on torch/ao/sparsity/_experimental/activation_sparsifier/README.md:
````diff
@@ -0,0 +1,75 @@
+2. **Reduce**: This is initiated once the ```step()``` is called. The ```reduce_fn()``` is called on the aggregated tensor.
+3. **Masking**: The computation of the mask happens immediately after the reduce operation. The ```mask_fn()``` is applied on the reduced tensor. Again, this happens per-feature, if the features are specified.
+
+## API Design
````
Please, add the description of the signatures for the agg, red, and mask functions. If I were to write those functions, what should I expect as inputs? What should a valid output be?

macandro96 replied:
Good point. Done.

z-a-f commented on torch/ao/sparsity/_experimental/activation_sparsifier/README.md:
```diff
@@ -0,0 +1,75 @@
+    return torch.eye(data.shape).to(data.device)
+
+# define activation sparsifier
+act_sparsifier = ActivationSparsifier(model, aggregate_fn, reduce_fn, mask_fn)
```
nit: I am not sure if PyTorch supports enforcing the "keyword args only", but I think we should at least *encourage* passing the functions using named args.

Commit:
Update on "[ao][sparsity] README for activation sparsifier"


The README contains introduction and details on the activation sparsifier. It also contains
code snippets and examples on using the activation sparsifier.

Test Plan: None

[ghstack-poisoned]

Commit:
Update on "[ao][sparsity] README for activation sparsifier"


The README contains introduction and details on the activation sparsifier. It also contains
code snippets and examples on using the activation sparsifier.

Test Plan: None

[ghstack-poisoned]

macandro96 commented:
@pytorchbot merge -g
